File: Python/Permutations.py
# Given a collection of numbers, return all possible permutations.

# For example,
# [1,2,3] have the following permutations:
# [1,2,3], [1,3,2], [2,1,3], [2,3,1], [3,1,2], and [3,2,1].


class Solution:
    # @param num, a list of integer
    # @return a list of lists of integers
	def permute(self, num):
		result = []
		visited = [0 for i in range(0,len(num))]
		solution=[]
		step = 0
		if len(num) == 0:
			return []
		self.findpermute(visited,solution,result,num,step)
		return result
	def findpermute(self,visited,solution,result,num,step):
		if step == len(num):
			result.append(list(solution))
			return
		for i in range(0,len(num)):
			if visited[i] == 0:
				visited[i] = 1
				solution.append(num[i])
				self.findpermute(visited,solution,result,num,step+1)
				solution.pop()
				visited[i]=0

File: Java/Permutations.java
class Permutations{
	public ArrayList<ArrayList<Integer>> permute(int[] num) {
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        ArrayList<Integer> sol = new ArrayList<Integer>();
        if(num == null || num.length == 0)
        	return res;
        int[] visited = new int[num.length];
        genPermute(num,visited,sol,res);
        return res;
    }
    private void genPermute(int[] num, int[] visited, ArrayList<Integer> sol, ArrayList<ArrayList<Integer>> res){
    	if(sol.size() == num.length){
    		res.add(new ArrayList<Integer>(sol));
    		return;
    	}
    	for(int i=0;i<num.length;i++){
    		if(visited[i] == 0){
    			visited[i] = 1;
    			sol.add(num[i]);
    			genPermute(num,visited,sol,res);
    			visited[i] = 0;
    			sol.remove(sol.size()-1);
    		}
    	}
    }
}